az-cvm-vtpm/src/vtpm/mod.rs

```diff
@@ -88,16 +92,80 @@ pub enum ReportError {
 
 /// Get a HCL report from an nvindex
 pub fn get_report() -> Result<Vec<u8>, ReportError> {
+    let (nv_index, mut context) = get_session_context()?;
+
+    let report = nv::read_full(&mut context, NvAuth::Owner, nv_index)?;
+    Ok(report)
+}
+
+/// Get a HCL report from an nvindex using temporary nvindex and report_data which will trigger
+/// refresh on main nvindex. Given report_data will be stored in HCL report in user_data field
+pub fn get_report_with_data(report_data: &[u8]) -> Result<Vec<u8>, ReportError> {
+    let (nv_index, mut context) = get_session_context()?;
+
+    let nv_index_temp = NvIndexTpmHandle::new(VTPM_HCL_REPORT_NV_INDEX_TEMP)?;
+    write_nv_index(&mut context, nv_index_temp, report_data);
+
+    // Adding a sleep for 3s for the user data to be reflected in VTPM_HCL_REPORT_NV_INDEX
+    thread::sleep(core::time::Duration::from_secs(3));
+
+    let report = nv::read_full(&mut context, NvAuth::Owner, nv_index)?;
+    Ok(report)
+}
+
+fn get_session_context() -> Result<(NvIndexTpmHandle, Context), ReportError> {
     use tss_esapi::handles::NvIndexTpmHandle;
     let nv_index = NvIndexTpmHandle::new(VTPM_HCL_REPORT_NV_INDEX)?;
 
     let conf: TctiNameConf = TctiNameConf::Device(DeviceConfig::default());
     let mut context = Context::new(conf)?;
     let auth_session = AuthSession::Password;
     context.set_sessions((Some(auth_session), None, None));
+    Ok((nv_index, context))
+}
 
-    let report = nv::read_full(&mut context, NvAuth::Owner, nv_index)?;
-    Ok(report)
+fn write_nv_index(context: &mut Context, nv_index: NvIndexTpmHandle, data: &[u8]) -> () {
+    let owner_nv_index_attributes = NvIndexAttributesBuilder::new()
+        .with_owner_write(true)
+        .with_owner_read(true)
+        .with_pp_read(true)
+        .build()
+        .expect("Failed to create owner nv index attributes");
+
+    let owner_nv_public = NvPublicBuilder::new()
+        .with_nv_index(nv_index)
+        .with_index_name_algorithm(HashingAlgorithm::Sha256)
+        .with_index_attributes(owner_nv_index_attributes)
+        .with_data_area_size(data.len())
+        .build()
+        .unwrap();
+
+    let exists = nv::list(context)
+        .unwrap()
+        .iter()
+        .map(|(public, _)| public.nv_index())
+        .any(|x| x == nv_index);
+
+    let mut rw;
+    if !exists {
+        rw = nv::NvOpenOptions::NewIndex {
+            nv_public: owner_nv_public,
+            auth_handle: NvAuth::Owner,
+        }
+            .open(context)
+            .unwrap();
+    } else {
+        rw = nv::NvOpenOptions::ExistingIndex {
+            nv_index_handle: nv_index,
+            auth_handle: NvAuth::Owner,
+        }
+            .open(context)
+            .unwrap();
+    }
+
+    rw.write_all(&data).unwrap();
+
+    drop(rw);
```

Comment:
why do we have to drop the handle explicitly here? shouldn't it be dropped automatically as it goes out of scope?

Reply:
You are right - it was previously added to use again context in same method